### ISSUE TITLE
Hack require in Webpack compilation

### ISSUE THREAD

**Issue by dtothefp** [halt-hammerzeit/webpack-isomorphic-tools#42]
I'm wondering if you ever discovered a way/created a plugin for Webpack to redefine requires for server compilation through Webpack. You suggested here https://github.com/halt-hammerzeit/webpack-isomorphic-tools/issues/7#issuecomment-149092151 the maybe this could be done through `NormalModuleReplacement` plugin. 

`webpack-isomorphic-tools` are working great for me if I pre-compile my files with Babel or use babel-register but I'm running into situations, such as on AWS Lambda, that I would prefer to make a server bundle with Webpack, then I don't have to zip all my client node_modules and upload them to Lambda.

Wondering if there is an easy way to do this?

**Comment by halt-hammerzeit** [comment]
I see, so you're talking about code duplication: you have your client build with `node_modules` merged into it and also you have a separate `node_modules` folder for server side.

I think that currently webpack doesn't have means to resolve such code duplication.

If you find Webpack bundles smaller than `node_modules` folder then you may build your server-side code too with `target: 'node'`
Like this
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example

**Comment by dtothefp** [comment]
yes I'm talking about building with `target: node`. With AWS Lambda you have to zip all the node_modules your lamda function needs to use. So for example if I want to build my React HTML string based upon some context/payload that will be passed from the lamda currently I'm 
- precompiling all my React component files with Babel
- uploading all of these component files + uploading all Node modules that the React components will utilize
- running webpack isomporphic tools through the lambda process and in turn creating my React HTML String

It would be nice to before uploading to the Lambda
- precompible/bundle all of my component files with webpack. This way they will also bundle the client libs such as React, etc that are `require`d in the component tree
- now I don't have to upload all those client node_modules to the lamda
- therefore it would be nice if there were a way to "hack" the webpack requires of the the relative file paths that are parsed in the webpack stats by isomorphic tools so when I run isomorphic tools on the Lamda it just works on my precompiled bundle rather than a tree of mutliple files.

Therefore, I would have run my client builds already to generate the stats through isomorphic tools, and I would create a second bundle with `target: node`

**Comment by halt-hammerzeit** [comment]
Ok. Well, if you think `target: 'node'` suits your needs then it's simple to implement.
You can either use the official example:
https://github.com/webpack/react-starter
Or you can use my example:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example

Also if you are planning on excluding `node_modules` from the build then you should also disable `externals`.
Respectively:
Here:
https://github.com/webpack/react-starter/blob/master/make-webpack-config.js#L78-L83
Or here:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example/blob/master/webpack/production%20build.js#L100

**Comment by dtothefp** [comment]
Yeh, I've excluded `node_modules` before so know how to do that. I was having a look at your example https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example trying to pick through it to see where the magic happens for the server builds and redefining of `require` for assets. Is there a specific file or two I should look at in that example?

**Comment by halt-hammerzeit** [comment]
No!
It doesn't redefine require()s at all because it doesn't need to.

**Comment by dtothefp** [comment]
ahhh...ok. Assuming the magic is happening here https://github.com/halt-hammerzeit/react-isomorphic-render?

**Comment by halt-hammerzeit** [comment]
There's no `magic` actually.
It's simple.

**Comment by dtothefp** [comment]
Hmmm...I don't seem to understand completely. I'm mostly wondering how I could make my server bundle where I 

```js
const styles = require('some/styles.scss');
const src = require('some/image.jpg');
```

work in webpack for server bundle.  I looked through 
https://github.com/halt-hammerzeit/react-isomorphic-render 
and
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example
and don't see anywhere that you are actually `require`ing css/scss inside your components, other than the `require.include` at the beginning of the `react-isomorphic-render/redux` `render` function, which I don't completely understand what that is doing?? 

Also, saw component level local scss with template strings and https://github.com/halt-hammerzeit/react-styling. 

Seems to be a lot of moving parts and I'm not really clear yet on how they all fit together?

I guess if I'm already generating stats from client bundles with `webpack-isomorphic-tools` I can just redefine the sass/file loaders for the server bundle to just supply info from the stats, but I'm assuming that is part of what the previous modules I listed are trying to solve, of not having to wait on the client stats/compilation in order to compile the server bundle with Webpack?

**Comment by halt-hammerzeit** [comment]
You just write 
```
const styles = require('some/styles.scss');
const src = require('some/image.jpg');
```

In your components and it should work.
For example, you can clone my project and run it to verify that it works initially, then insert those lines in some component and check if it works. You can experiment with anything.

I'm not requiring styles in my components because i'm using inline styles. Still there's nothing preventing styles and images from being required in the code.

`render` function just renders Html pages on the server and React components on the client.

Also I've updated the readme. Have a look:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example#brief-explanation

**Comment by danecando** [comment]
I'm experiencing this issue as well and I think there's some confusion. Any of the examples that I've seen do not generate node server bundles with webpack, they use a babel require hook. 

So when webpack is actually parsing my node application to generate a bundle, I get an error during the compilation process about my requires. I'm not actually starting a node server to run my app, just bundling it with webpack causes the error.

Has anyone have any more insight on this?

**Comment by halt-hammerzeit** [comment]
@danecando "Anyone" is "me" to start with.
What "issue" are you experiencing?
There's no "issue" here.
You want webpack compiled server side bundle? You do it.
As simple as that.
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example
And in that case you won't need `webpack-isomorphic-tools`.

**Comment by dtothefp** [comment]
I ended up still using isomorphic tools and make my server bundle with webpack after I generate my client bundles. Then I can just use a mock loader that I created for webpack that when it sees styles or images...whatever is breaking your build it is in the isomorphic tools stats, I just read the stats and objects the data from stats. Works well for me because I don't need my client bundle and server bundle to be generated concurrently because I'm generating static markup not using an application server to generate html content. Also if you mark all your commonjs modules as external for the s Tver build it won't bundle them and your build will be relatively fast. 

I think what @halt-hammerzeit is solving in https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example is how to generate his client and server bundle with Webpack concurrently. It looks awesome but I haven't had time to parse through and understand the logic. On the next app I'm building where I need an application server I hope to get back to it. 

Thanks @halt-hammerzeit for adding more docs to the README for your demo btw :wink:

**Comment by danecando** [comment]
@halt-hammerzeit I was referring to a few different boilerplates / examples that I've seen, not specifically you. I was simply asking a question about using it for webpack compiled server bundles but wanted to make sure my use case was clear so I could be pointed in the right direction.  No need to get in a tiff. Suggestions: coffee, maybe a nice black or green tea before you get on GitHub in the morning.. Maybe even some meditation! I haven't yet seen that repo / example but that's pretty much what I'm looking to accomplish. Keep up the good work, it's much appreciated. (sorry if you took my previous question the wrong way)

**Comment by halt-hammerzeit** [comment]
@dtothefp oh, I see. yes, that sample project of mine generates client and server builds concurrently in development. in production though it does it sequentially.

@danecando no, it's just funny to rant on random people on the internets so GET YOUR SHIT TOGETHER.

**Comment by halt-hammerzeit** [comment]
@danecando 
> So when webpack is actually parsing my node application to generate a bundle, I get an error during the compilation process about my requires. I'm not actually starting a node server to run my app, just bundling it with webpack causes the error.

You shouldn't get any errors about your requres because Webpack knows how to compile and bundle all the shit together.
Usually if you are having errors of some kind it's kinda useful to provide a thing called "an error message".
That helps, you know.

**Comment by danecando** [comment]
@halt-hammerzeit not sure I'm getting you, but I like where your head is at. I should have slept on my issue, it was two hours past my bed time and I was just trying to finish getting this tool implemented in my project. So I jumped the gun a little bit there. I will wait till I have taken the time to really dig into this before bothering you again, if necessary. We can still take jabs at each other though if you'd like, I'm always game for that.

**Comment by halt-hammerzeit** [comment]
@danecando I take fun as a payment for helping people resolve THEIR issues. You think that's unfair? It's your decision. Only yours.

**Comment by danecando** [comment]
It depends... are you harvesting all this fun for some sort of diabolical plan to make the world a funner place? Because that I just can't get on board with.

**Comment by halt-hammerzeit** [comment]
@danecando You can't get on board with everything. That's how this world works. Deal with it.

**Comment by danecando** [comment]
maybe in mother russia nikoliteral

**Comment by halt-hammerzeit** [comment]
@danecando Mother Russia is the Divine Core of rebuilding the Human Civilization when the world falls into chaos. Soon.

**Comment by danecando** [comment]
the power of christ compels you

**Comment by halt-hammerzeit** [comment]
@danecando I'm not a Christian. I believe in far more complicated things. Like the Overmind.

**Comment by halt-hammerzeit** [comment]
@danecando There are a lot of christians here though. We don't fight with each other yet.
http://politicalhotwire.com/current-events/119136-obamas-lack-education-display-again-time-equating-christianity-isis-15.html

![xxx](http://img.theepochtimes.com/n3/eet-content/uploads/2014/11/10/AP335212256729.jpg)

**Comment by danecando** [comment]
I know. I was performing an e-xorcism to eradicate the spirit of empress anna from your human body.

**Comment by halt-hammerzeit** [comment]
@danecando You are talking nonsense. I think you're kinda insane. Oh, and American society is spiritless.

Now, I'm gonna go watch some armpit licking lesbian porn so I'm no longer interested in this pointless discussion with you.

**Comment by halt-hammerzeit** [comment]
FYI
I'm posting this in every issue and PR to notify whoever may be interested:
today I've released an alternative helper library called [`universal-webpack`](https://github.com/halt-hammerzeit/universal-webpack).
It takes a different approach than `webpack-ismorphic-tools` and instead of hacking Node.js `require()` calls it just compiles all code with `target: 'node'` webpack configuration option.
As a result, all Webpack plugins and features are supported.
If you think you might need that here's an example project:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example
